File: week2/LongestPalindromicSubstring/longestPalindromicSubstring.cpp
#include <iostream>
using namespace std;

std::string longestPalindromicSubstring(std::string str);
bool isPalindromic(std::string s);

int main(){
    // Input: s = "babad"  Output: "bab"
    std::string input ="babad";
    std::string output="";
    output = longestPalindromicSubstring(input);
    cout << output;
}

bool isPalindromic(std::string s){
    if (s.length()==1){
        return true;
    }
    int left=0, right=s.length()-1;
    while(left<right){
        if(s[left] != s[right]){
            return false;
        }
        left++;
        right--;
    }
    return true;
}

std::string longestPalindromicSubstring(std::string str){
    if (str.length()<=1){
        return str;
    }
    int max_len=1;
    std::string max_str=str.substr(0,1);
    for(int i=0; i<str.length();i++){
        for(int j=i;j<str.length();j++){
            if (j-i > max_len && isPalindromic(str.substr(i,j))){
                max_len = j-i;
                max_str = str.substr(i,j);
            }
        }
    }
    return max_str;
}

File: week2/LongestPalindromicSubstring/longestPalindromicSubstring.py
class Solution:
    @classmethod
    def isPalindromic(self, s: str)-> bool:
        left=0
        right=len(s)-1
        while left<right:
            if s[left] != s[right]:
                return False
            left+=1
            right-=1
        return True

    @classmethod
    def longestPalindromicSubstring(self, input: str)->str:
        if len(input) <= 1:
            return input
        max_len=1
        max_str=input[0]
        for i in range(len(input)):
            for j in range(i+max_len, len(input)):
                if j-i > max_len and self.isPalindromic(input[i:j]):
                    max_len = j-i
                    max_str = input[i:j]
        return max_str

# Input: s = "babad"
# Output: "bab"
input = "babad"
output = Solution.longestPalindromicSubstring(input)
print(output)